### ISSUE TITLE
OpenAPI loading double-buffering (or triple-buffering) 

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#119983]
### What would you like to be added?

While inspecting the code for https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go, it appears that we buffer the OpenAPI response _twice_ as a []byte, once in io.ReadAll when we are reading it for parsing, and once in the HTTP caching layer.

Arguably we need 3x the memory because our json parser is not streaming, so we retain the []byte while we unmarshal into the OpenAPI types.

Given we are writing it to disk anyway, we should probably just write to a temp file.

### Why is this needed?

This is inefficient, and is visible in memory profiles.

I think we can also eliminate two library dependencies from our stack (possibly even 3).: https://github.com/gregjones/httpcache is archived.  diskv has not had many updates in years.  diskv pulls in google/btree, so we might also be able to eliminate that.

**Comment by pegasas** [comment]
I wants to follow this issue.
/assign @pegasas

**Comment by pegasas** [comment]
Hi, @justinsb ,

looks like we no longer need [round_tripper](https://github.com/kubernetes/client-go/blob/master/discovery/cached/disk/round_tripper.go) and only keeps [cached_discovery](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go) as cache layer?

**Comment by apelisse** [comment]
You probably don't want to be assigned but CC, which you already are. 

> looks like we no longer need [round_tripper](https://github.com/kubernetes/client-go/blob/master/discovery/cached/disk/round_tripper.go) and only keeps [cached_discovery](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go) as cache layer?

What makes you think that?

**Comment by pegasas** [comment]
Hi, @apelisse

For me, I would like to follow this issue and work on a PR on it.
Clearly, You are the expert and know better than me. 
If you want to follow your own, that's OK.

> What makes you think that?

> While inspecting the code for https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go, it appears that we buffer the OpenAPI response _twice_ as a []byte, once in io.ReadAll when we are reading it for parsing, and once in the HTTP caching layer.
> 
> Arguably we need 3x the memory because our json parser is not streaming, so we retain the []byte while we unmarshal into the OpenAPI types.
> 
> Given we are writing it to disk anyway, we should probably just write to a temp file.

I read @justinsb 's solution and found [round_tripper](https://github.com/kubernetes/client-go/blob/master/discovery/cached/disk/round_tripper.go) as HTTP caching layer. 

@apelisse Is this solution missing some important architectural thinking there?

**Comment by apelisse** [comment]
Yeah, you need the round-tripper. From my read in this, the most reasonable fix is to:
1. Find out if gregones/httpcache can be updated or not (it's been archived so unlikely), decide if we want to fork it and how
2. Update the cache saving mechanism to just write the data as it's read (instead of saving into a buffer), inside a temporary file (has to be in on the same disk though so that you can ...), rename the file once the read is complete.

The problematic line is probably [here](https://github.com/gregjones/httpcache/blob/master/httpcache.go#L535C20-L535C20), it's writing to a buffer which forces the double allocation.

To be honest, you would have to change the `cachingReadCloser`, change the `OnEOF` callback to do what you want (dancing around the `httputil.DumpResponse`), and figure out a way to workaround the `t.Cache.Set` API which, I would say, you can change under no circumstances.

It sounds pretty hard tbh.

**Comment by pegasas** [comment]
> Yeah, you need the round-tripper. From my read in this, the most reasonable fix is to:
> 
> 1. Find out if gregones/httpcache can be updated or not (it's been archived so unlikely), decide if we want to fork it and how
> 2. Update the cache saving mechanism to just write the data as it's read (instead of saving into a buffer), inside a temporary file (has to be in on the same disk though so that you can ...), rename the file once the read is complete.
> 
> The problematic line is probably [here](https://github.com/gregjones/httpcache/blob/master/httpcache.go#L535C20-L535C20), it's writing to a buffer which forces the double allocation.
> 
> To be honest, you would have to change the `cachingReadCloser`, change the `OnEOF` callback to do what you want (dancing around the `httputil.DumpResponse`), and figure out a way to workaround the `t.Cache.Set` API which, I would say, you can change under no circumstances.
> 
> It sounds pretty hard tbh.

This route is OK.
I would like to add an `AppendStream` method into diskv package.
So that we could directly append by sumDiskCache.

Here's the issue I am tracking. https://github.com/peterbourgon/diskv/issues/77

Feel free to comments if any concern or better idea.

**Comment by apelisse** [comment]
Doesn't sound like this is going down as you expected. Even if you did change diskv, I think you would still have to change gregjones Cache interface, which is not an option. I'm not necessarily convinced this bug is a priority to fix though.
